text/0059-otlp-trace-data-format.md

```diff
@@ -0,0 +1,349 @@
+  //     "/http/server_latency": 300
+  //     "abc.com/myattribute": true
+  //     "abc.com/score": 10.239
+  repeated AttributeKeyValue attributes = 10;
+
+  // The number of attributes that were discarded. Attributes can be discarded
+  // because their keys are too long or because there are too many attributes.
+  // If this value is 0, then no attributes were dropped.
+  int32 dropped_attributes_count = 11;
+
+  // TimedEvent is a time-stamped annotation of the span, consisting of either
+  // user-supplied key-value pairs, or details of a message sent/received between Spans.
+  message TimedEvent {
+    // The time the event occurred.
+    int64 time_unixnano = 1;
+
+    // A user-supplied name describing the event.
+    string name = 2;
+
+    // A list of attributes of the event.
+    repeated AttributeKeyValue attributes = 3;
+
+    // The number of dropped attributes. If the value is 0, then no attributes were dropped.
+    int32 dropped_attributes_count = 4;
+  }
+
+  // timed_events is a collection of `TimedEvent`s.
+  repeated TimedEvent timed_events = 12;
+
+  // dropped_timed_events_count is the number of dropped timed events. If the value is 0, then
+  // no events were dropped.
+  int32 dropped_timed_events_count = 13;
+
+  // Link is a pointer from the current span to another span in the same trace or in a
+  // different trace. For example, this can be used in batching operations,
+  // where a single batch handler processes multiple requests from different
+  // traces or when the handler receives a request from a different project.
+  message Link {
+    // A unique identifier of a trace that this linked span is part of. The ID is a
+    // 16-byte array.
+    bytes trace_id = 1;
+
+    // A unique identifier for the linked span. The ID is an 8-byte array.
+    bytes span_id = 2;
+
+    // The Tracestate associated with the link.
+    repeated TraceEntry tracestate = 3;
+
+    // A list of attributes of the link.
+    repeated AttributeKeyValue attributes = 4;
+
+    // dropped_attributes_count is the number of dropped attributes. If the value is 0, then
+    // no attributes were dropped.
+    int32 dropped_attributes_count = 5;
+  }
+
+  // links is a collection of Links, which are references from this span to a span
+  // in the same or different trace.
+  repeated Link links = 14;
+
+  // dropped_links_count is the number of dropped links after the maximum size was enforced.
+  // If this value is 0, then no links were dropped.
+  int32 dropped_links_count = 15;
+
+  // An optional final status for this span. Semantically when Status
+  // wasn't set it is means span ended without errors and assume
+  // Status.Ok (code = 0).
+  Status status = 16;
+
+  // An optional number of child spans that were generated while this span
+  // was active. If set, allows an implementation to detect missing child spans.
+  int32 child_span_count = 17;
```

Comment:
As I read the description, this will only count "local children". I think that it would be even more valuable for the back end to know the number of potential "remote children", i.e. the number of times the Span's SpanContext was propagated (this is also one of the motivations for #58 "Unseal SpanContext", but if that was doable with the reference SDK, that would of course be even nicer).

Reply:
Does the number of times SpanContext was propagated necessarily equal to the number of children created elsewhere? You may propagate once and create miltiple children, or propagate but do not create any children at all. I do not see how a number of propagations can reflect number of children or how you would otherwise be able to know how many child spans are generated somewhere else that you don't control.

Reply:
Yes it's true that this is only a "potential" number of children, and not even a maximum. After reading your arguments, I guess the only value in a propagation count would be to check if it is > 0. So maybe a `bool was_propagated` would be good?

Reply:
@Oberon00 you can use a combination of Kind + this to determine if the number is correctly calculated or not. For the Kind Server/Consumer/Internal the number is correct, for the Client/Producer the number will be 0 but without having a response header that determines if the next service created or not a Span is hard (even harder for Producer where the Consumer may be async and this Span will be sent ended).

@tigrannajaryan: in opencensus this was an Uint32 message because we believed that having the ability to use not-set value was important for cases like Client/Producer. What do you think?

Reply:
@bogdandrutu I feel you assign more meaning to SpanKind than it reallly has. Cases where your assumption breaks:
* There is a DB client span that has a HTTP client span as local child that implements executing the request.
* The client call fails before an attempt to propagate the context is even made (e.g. connection refused for HTTP request).
* A client call is made but propagation is explicitly disabled (for example an HTTP integration may offer an option to suppress propagation for URLs that are known to go to external services).

Reply:
>in opencensus this was an Uint32 message because we believed that having the ability to use not-set value was important for cases like Client/Producer. What do you think?

@bogdandrutu yes, 0 in this case semantically is different from "not set" (which would mean we don't know). However I have a hard time imagining a use case where this difference matters.
In my mind I can imagine one good use case for `child_span_count` field: to indicate that there are missing children. For this I want to know that `child_span_count>0` and perhaps also show the actual counter in the UI (just to look nice). If the number is 0 or "not set" I will likely interpret it the same way, which is "I am not going to do anything special in this case".
Perhaps there is a good use case where this difference matters, if you have examples that would be great to know.

Reply:
I am resolving this since there are no further objections.